### PR TITLE
fix datetime format

### LINKED ISSUE #4487: invalid format for datetime
```
(root@nebula) [sf100]> return datetime("1659602415")
+------------------------------+
| datetime("1659602415")       |
+------------------------------+
| -31249-01-01T00:00:00.000000 |
+------------------------------+
Got 1 rows (time spent 298/12422 us)

Thu, 04 Aug 2022 16:52:09 CST
```

we support datetime string, e.g. "2017-03-04T22:30:40.003000[Asia/Shanghai]"
and "1659602415" should be prevented.

### PR DESCRIPTION
**Pull request by caton-hpg**
## What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### Description:
fixed #4487.

It will be `BAD_DATA` if input a bad year or month.
```
(root@nebula) [(none)]> return datetime("32767");
+-----------------------------+
| datetime("32767")           |
+-----------------------------+
| 32767-01-01T00:00:00.000000 |
+-----------------------------+
Got 1 rows (time spent 1900/2890 us)


(root@nebula) [(none)]> return datetime("32768");
+-------------------+
| datetime("32768") |
+-------------------+
| BAD_DATA          |
+-------------------+
Got 1 rows (time spent 1265/1800 us)


(root@nebula) [(none)]> return datetime("-32768");
+------------------------------+
| datetime("-32768")           |
+------------------------------+
| -32768-01-01T00:00:00.000000 |
+------------------------------+
Got 1 rows (time spent 1127/1800 us)


(root@nebula) [(none)]> return datetime("-32769");
+--------------------+
| datetime("-32769") |
+--------------------+
| BAD_DATA           |
+--------------------+
Got 1 rows (time spent 1084/1699 us)


```

## How do you solve it?
year is in [-32768,32767].

## Checklist:
Tests:
- [x] Unit test(positive and negative cases)
- [x] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:

### CODE REVIEW

**Comment by Shylock-Hg** [comment]
I think is a out of range date, so you could check range and return error. Don't modify parser (and test).

**Comment by caton-hpg** [comment]
Fixed.

year is in [0, 9999]
month is in [1, 12]

File: src/common/time/TimeUtils.h [hunk 1]
```diff
@@ -44,6 +44,14 @@ class TimeUtils {
     if ((p[date.month] - p[date.month - 1]) < date.day) {
       return Status::Error("`%s' is not a valid date.", date.toString().c_str());
     }
+    if (date.year < 0 || date.year > 9999) {
```
Comment (Shylock-Hg) [inline src/common/time/TimeUtils.h:47]: Check it in parser, greater than `9999` is also valid, our upper bound is `int32_t::max()`.

Comment (Shylock-Hg) [reply src/common/time/TimeUtils.h:47]: Lower bound too.

Comment (caton-hpg) [reply src/common/time/TimeUtils.h:47]: I talk with @HarrisChu . 
He thinks `date.year` should be year. Should not be `int32_t::max()`.

Comment (Shylock-Hg) [reply src/common/time/TimeUtils.h:47]: It's obviously incorrect, we should provide time range as wide as possible for user. You could check Postgres in https://www.postgresql.org/docs/9.1/datatype-datetime.html or other databases.

Comment (HarrisChu) [reply src/common/time/TimeUtils.h:47]: > 

what's your expected for below queries:

```
yield datetime(-1)
yield datetime(1000)
yield datetime(123456)

yield datetime("-1")
yield datetime("1000")
yield datetime("123456")
```

Comment (Shylock-Hg) [reply src/common/time/TimeUtils.h:47]: All valid date.

Comment (Shylock-Hg) [reply src/common/time/TimeUtils.h:47]: Negative date means B.C.

Comment (HarrisChu) [reply src/common/time/TimeUtils.h:47]: image

one is for epoch, the other is just year

Comment (Shylock-Hg) [reply src/common/time/TimeUtils.h:47]: datetime("-1") is unexpected.

Comment (caton-hpg) [reply src/common/time/TimeUtils.h]: fixed. year is in [-32767,32767].

File: src/common/time/TimeUtils.cpp [hunk 2]
```diff
@@ -25,7 +25,8 @@ constexpr int64_t kMaxTimestamp = std::numeric_limits<int64_t>::max() / 10000000
       return Status::Error("Invalid value type.");
     }
     if (kv.first == "year") {
-      if (kv.second.getInt() < std::numeric_limits<int16_t>::min() ||
+      // year should be in [-32767, 32767] and same as parser
+      if (kv.second.getInt() <= std::numeric_limits<int16_t>::min() ||
```
Comment (Shylock-Hg) [inline src/common/time/TimeUtils.cpp:29]: Why less and equal?

Comment (caton-hpg) [reply src/common/time/TimeUtils.cpp:29]: The max of year in paser is 32767. The min is -32767.
In order to be same as parser.

Comment (Shylock-Hg) [reply src/common/time/TimeUtils.cpp:29]: Where limit this in parser?

Comment (caton-hpg) [reply src/common/time/TimeUtils.cpp:29]: The year, month, day, hour and others are INTEGER.
INTEGER is a `[0-9]+` and cannot be a negative int.
Negative year supports by 
```
| NEGATIVE INTEGER NEGATIVE INTEGER NEGATIVE INTEGER {
    $$ = new nebula::Date(0-$2, $4, $6);
```

Comment (Shylock-Hg) [reply src/common/time/TimeUtils.cpp:29]: I can't catch you.

Comment (caton-hpg) [reply src/common/time/TimeUtils.cpp]: fixed. year is in [-32768,32767].

File: src/common/time/parser/test/DateTimeParserTest.cpp [hunk 3]
```diff
@@ -230,8 +259,24 @@ TEST(DatetimeReader, DateFailed) {
   // not exits prefix
   {
     auto parser = time::DatetimeReader();
-    auto result = parser.readDate("-01-03");
```
Comment (Shylock-Hg) [inline src/common/time/parser/test/DateTimeParserTest.cpp:233]: Don't modify origin test. Just add yours.

File: src/common/time/parser/test/DateTimeParserTest.cpp [hunk 4]
```diff
@@ -152,8 +159,24 @@ TEST(DatetimeReader, DateTimeFailed) {
   // not exits prefix
   {
     auto parser = time::DatetimeReader();
-    auto result = parser.readDatetime("-01-03T22:22:3.2333");
```
Comment (Shylock-Hg) [inline src/common/time/parser/test/DateTimeParserTest.cpp:155]: ditto

**Review by Shylock-Hg — DISMISSED** [review]

**Comment by Shylock-Hg** [comment]
We'd better avoid force-push

**Review by Shylock-Hg — APPROVED** [review]

**Review by codesigner — APPROVED** [review]
Nice

**Pull request merged**